### PR TITLE
bpo-30674: IDLE: add docstrings to grep.py

### PR DESCRIPTION
**Pull request by csabella**

### CODE REVIEW

File: Lib/idlelib/grep.py [hunk 1]
```diff
@@ -20,19 +35,32 @@ def grep(text, io=None, flist=None):
     searchphrase = text.get("sel.first", "sel.last")
     dialog.open(text, searchphrase, io)
 
+
 class GrepDialog(SearchDialogBase):
+    "Search Dialog for Find in Files."
```
Comment (mlouielu) [inline Lib/idlelib/grep.py]: `"""Search dialog for find in files."""`

File: Lib/idlelib/grep.py [hunk 2]
```diff
@@ -20,19 +35,32 @@ def grep(text, io=None, flist=None):
...
 
     title = "Find in Files Dialog"
     icon = "Grep"
     needwrapbutton = 0
 
     def __init__(self, root, engine, flist):
+        """Create Search Dialog for searching for a phrase in the file system.
```
Comment (mlouielu) [inline Lib/idlelib/grep.py]: `"""Create search dialog for ...`

File: Lib/idlelib/grep.py [hunk 3]
```diff
@@ -20,19 +35,32 @@ def grep(text, io=None, flist=None):
...
+            recvar: Boolean value of Checkbutton widget
+                    for traversing through subdirectories.
+        """
         SearchDialogBase.__init__(self, root, engine)
         self.flist = flist
         self.globvar = StringVar(root)
         self.recvar = BooleanVar(root)
 
     def open(self, text, searchphrase, io=None):
+        "Make dialog visible on top of others and ready to use."
```
Comment (mlouielu) [inline Lib/idlelib/grep.py:63]: `"""Make dialog visible .... """`

File: Lib/idlelib/grep.py [hunk 4]
```diff
@@ -45,36 +73,54 @@ def open(self, text, searchphrase, io=None):
         self.globvar.set(os.path.join(dir, "*" + tail))
 
     def create_entries(self):
+        "Create base entry widgets and add widget for search path."
```
Comment (mlouielu) [inline Lib/idlelib/grep.py:76]: `"""Create base entry .... path."""`

File: Lib/idlelib/grep.py [hunk 5]
```diff
@@ -45,36 +73,54 @@ def open(self, text, searchphrase, io=None):
...
         SearchDialogBase.create_entries(self)
         self.globent = self.make_entry("In files:", self.globvar)[0]
 
     def create_other_buttons(self):
+        "Add check button to recurse down subdirectories."
```
Comment (mlouielu) [inline Lib/idlelib/grep.py:81]: `"""Add check button ....subdirectories."""`

File: Lib/idlelib/grep.py [hunk 6]
```diff
@@ -45,36 +73,54 @@ def open(self, text, searchphrase, io=None):
...
         btn = Checkbutton(
                 self.make_frame()[0], variable=self.recvar,
                 text="Recurse down subdirectories")
         btn.pack(side="top", fill="both")
 
     def create_command_buttons(self):
+        "Create base command buttons and add button for search."
```
Comment (mlouielu) [inline Lib/idlelib/grep.py:88]: `"""Create base command .... for search."""`

File: Lib/idlelib/grep.py [hunk 7]
```diff
@@ -45,36 +73,54 @@ def open(self, text, searchphrase, io=None):
...
+
+        The default command is bound to <Return>.
+
+        If entry values are populated, set OutputWindow as stdout
+        and perform search.
+        """
         prog = self.engine.getprog()
         if not prog:
-            return
+            return None
```
Comment (mlouielu) [inline Lib/idlelib/grep.py]: `return` will return `None`.  So you don't need to explicit `return None`

File: Lib/idlelib/grep.py [hunk 8]
```diff
@@ -124,9 +174,8 @@ def findfiles(self, dir, base, rec):
         return list
 
     def close(self, event=None):
-        if self.top:
-            self.top.grab_release()
-            self.top.withdraw()
+        "Close the dialog."
```
Comment (mlouielu) [inline Lib/idlelib/grep.py]: `"""Close the dialog."""`

**Review by mlouielu — CHANGES_REQUESTED** [review]
Thanks for the PR, some small point should be changed.

**Comment by mlouielu** [comment]
I also create a picture for the import relationship for IDLE.  you can saw it here:

**Comment by csabella** [comment]
Terry had changed my """ quotes on a previous PR to just " when the docstring was only one line.  I was following his convention with these changes.

File: Lib/idlelib/grep.py (hunk 1 above)
Comment (terryjreedy) [reply Lib/idlelib/grep.py]: The menu item is 'Find in Files', but I agree it is not obvious and awkward if one does not notice.  "Find in Files" is already in file docstring.  How about "Dialog for searching multiple files."?

File: Lib/idlelib/grep.py [hunk 9]
```diff
@@ -1,3 +1,8 @@
+"""Grep dialog for Find in File functionality.
```
Comment (terryjreedy) [inline Lib/idlelib/grep.py]: 'Files', plural

File: Lib/idlelib/grep.py [hunk 10]
```diff
@@ -11,7 +16,17 @@
...
 # EditorWindow -> GrepDialop -> OutputWindow -> EditorWindow
 
+
 def grep(text, io=None, flist=None):
+    """Create or find singleton GrepDialog instance.
+
+    Args:
+        text: Text widget that contains the selected text for
+              default search phrase.
+        io: IOBinding with default path to search.
```
Comment (terryjreedy) [inline Lib/idlelib/grep.py]: 'iomenu.IOBinding instance'

File: Lib/idlelib/grep.py (hunk 2 above)
Comment (terryjreedy) [reply Lib/idlelib/grep.py]: agreed

File: Lib/idlelib/grep.py (hunk 3 above)
Comment (terryjreedy) [reply Lib/idlelib/grep.py:63]: No. Single double quote is allowed and preferred by me.  Common in stdlib.

File: Lib/idlelib/grep.py [hunk 11]
```diff
@@ -45,36 +73,54 @@ def open(self, text, searchphrase, io=None):
...
     def create_command_buttons(self):
+        "Create base command buttons and add button for search."
         SearchDialogBase.create_command_buttons(self)
         self.make_button("Search Files", self.default_command, 1)
 
     def default_command(self, event=None):
+        """Grep for search pattern in file path.
+
+        The default command is bound to <Return>.
+
```
Comment (terryjreedy) [inline Lib/idlelib/grep.py]: Don't need new paragraph here.

File: Lib/idlelib/grep.py (hunk 7 above)
Comment (terryjreedy) [reply Lib/idlelib/grep.py]: "Either all return statements in a function should return an expression, or none of them should."  So if any must (return an expression), all must, but if all return None, then none must, so None can be omitted but must be omitted everywhere.

File: Lib/idlelib/grep.py [hunk 12]
```diff
@@ -45,36 +73,54 @@ def open(self, text, searchphrase, io=None):
...
+        """
         prog = self.engine.getprog()
         if not prog:
-            return
+            return None
         path = self.globvar.get()
         if not path:
             self.top.bell()
-            return
+            return None
```
Comment (terryjreedy) [inline Lib/idlelib/grep.py]: Leave this at it was also.

File: Lib/idlelib/grep.py (hunk 8 above)
Comment (terryjreedy) [reply Lib/idlelib/grep.py]: No.

File: Lib/idlelib/grep.py [hunk 13]
```diff
@@ -96,14 +142,18 @@ def grep_it(self, prog, path):
                 except OSError as msg:
                     print(msg)
             print(("Hits found: %s\n"
-                  "(Hint: right-click to open locations.)"
-                  % hits) if hits else "No hits.")
+                   "(Hint: right-click to open locations.)"
+                   % hits) if hits else "No hits.")
```
Comment (terryjreedy) [inline Lib/idlelib/grep.py]: I believe this was technically correct as it was, but it looks better with the extra space.  But I don't really like it either way (and I wrote it in this form).  In 3.6+, we can do
```
            print(f"Hits found: {hits}\n"
                  f"(Hint: right-click to open locations.)"
                  if hits else "No hits.")
            # or with 79 char lines
            print(f"Hits found: {hits}\n(Hint: right-click to open locations.)"
                  if hits else "No hits.")
```
I like both of these better, so choose one.

File: Lib/idlelib/grep.py [hunk 14]
```diff
@@ -124,9 +174,8 @@ def findfiles(self, dir, base, rec):
         return list
 
     def close(self, event=None):
-        if self.top:
-            self.top.grab_release()
-            self.top.withdraw()
+        "Close the dialog."
+        SearchDialogBase.close(self, event)
```
Comment (terryjreedy) [inline Lib/idlelib/grep.py]: Good catch.  Maybe these were once different.  Now, just delete GrepDialog.close and inherit SearchDialogBase.close.

File: Lib/idlelib/grep.py (hunk 4 above)
Comment (terryjreedy) [reply Lib/idlelib/grep.py:76]: See above.

File: Lib/idlelib/grep.py (hunk 5 above)
Comment (terryjreedy) [reply Lib/idlelib/grep.py:81]: Ditto.

File: Lib/idlelib/grep.py [hunk 15]
```diff
@@ -45,36 +73,54 @@ def open(self, text, searchphrase, io=None):
...
-            return
+            return None
         from idlelib.outwin import OutputWindow  # leave here!
         save = sys.stdout
         try:
             sys.stdout = OutputWindow(self.flist)
             self.grep_it(prog, path)
         finally:
             sys.stdout = save
+        return None
```
Comment (terryjreedy) [inline Lib/idlelib/grep.py]: " If any return statement returns an expression, any return statements where no value is returned should explicitly state this as return None "  So if no return statement return an expression, no explicit return is needed at the end.  In other words, this function was within the rules and should be left as it was.

**Review by terryjreedy — CHANGES_REQUESTED** [review]

File: Lib/idlelib/grep.py (hunk 14 above)
Comment (csabella) [reply Lib/idlelib/grep.py]: Thanks!

File: Lib/idlelib/grep.py (hunk 13 above)
Comment (csabella) [reply Lib/idlelib/grep.py]: Changed.  I also changed the other string formatting lines to use f strings.

File: Lib/idlelib/grep.py (hunk 15 above)
Comment (csabella) [reply Lib/idlelib/grep.py]: Got it.  Thanks.

**Comment by csabella** [comment]
@mlouielu  @terryjreedy    Thanks for the reviews.  I added an additional note to default_command to mention that the search dialog is closed when the search begins.

One other question -- when I run this, the dialog box always opens way on the bottom of my screen.  Has anyone requested a change to this before?  Or would it be worth mentioning on idle-dev?

**Comment by mlouielu** [comment]
@csabella I think your question should also post it on b.p.o. Since here is to discuss about code review. I didn't get this on macOS, the dialog show off at the left top side of the screen.

**Pull request merged**